go/database/mpt/verification_proof.go

```diff
@@ -0,0 +1,336 @@
+	}
+	if !complete {
+		return fmt.Errorf("proof incomplete for 0x%x", addr)
+	}
+	if got, want := nonce, data.nonce; got != want {
+		return fmt.Errorf("nonce mismatch for 0x%x, got %v, want %v", addr, got, want)
+	}
+	code, complete, err := proof.GetCodeHash(root, addr)
+	if err != nil {
+		return err
+	}
+	if !complete {
+		return fmt.Errorf("proof incomplete for 0x%x", addr)
+	}
+	if got, want := code, data.code; got != want {
+		return fmt.Errorf("code mismatch for 0x%x, got %v, want %v", addr, got, want)
+	}
+
+	return nil
+}
+
+// verifyStorageProof verifies the consistency between the input witness proofs and the storage.
+func verifyStorageProof(root common.Hash, proof witness.Proof, addr common.Address, keys []common.Key, storage map[common.Key]common.Value) error {
+	for _, key := range keys {
+		proofValue, complete, err := proof.GetState(root, addr, key)
+		if err != nil {
+			return err
+		}
+		if !complete {
+			return fmt.Errorf("proof incomplete for address: 0x%x, key: 0x%x", addr, key)
+		}
+		if got, want := proofValue, storage[key]; got != want {
+			return fmt.Errorf("storage mismatch for 0x%x, key 0x%x, got %v, want %v", addr, key, got, want)
+		}
+	}
+
+	return nil
+}
+
+// generateUnknownAddresses generates a slice of addresses that will never appear
+// in real MPT and the witness proof. Each address consists of four random
+// bytes, a number one, and the rest bytes are zero. The number one is set to
+// prevent generation of random address that has only zeros. Such an address
+// is reserved for a special contract.
+func generateUnknownAddresses(number int) []common.Address {
+	res := make([]common.Address, number)
+	for i := range res {
+		j := rand.Int()
+		res[i] = common.Address{byte(j), byte(j >> 8), byte(j >> 16), byte(j >> 24), 1}
+	}
+	return res
+}
+
+// generateUnknownKeys generates a slice of keys that will never appear in real
+// MPT and the witness proof. Each key consists of four random bytes, a number
+// one, and the rest bytes are zero. The number one is set to prevent generation
+// of random key that has only zeros. Such a key is reserved for a special contract.
+func generateUnknownKeys(number int) []common.Key {
+	res := make([]common.Key, number)
+	for i := range res {
+		j := rand.Int()
+		res[i] = common.Key{byte(j), byte(j >> 8), byte(j >> 16), byte(j >> 24), 1}
+	}
+
+	return res
+}
+
+type accountData struct {
+	balance amount.Amount
+	nonce   common.Nonce
+	code    common.Hash
+	storage map[common.Key]common.Value
+}
+
+func (a accountData) storageKeyRanges(rangeSize int) [][]common.Key {
+	keys := maps.Keys(a.storage)
+	var ranges [][]common.Key
+	for i := 0; i < len(keys); i += rangeSize {
+		end := i + rangeSize
+		// Ensure we don't exceed the slice bounds
+		if end > len(keys) {
+			end = len(keys)
+		}
+		ranges = append(ranges, keys[i:end])
+	}
+
+	return ranges
+}
+
+type accountsMap map[common.Address]accountData
+
+type accountCollectingVisitor struct {
+	accounts       accountsMap
```

Comment:
Collecting this full account map for a large state will require a large amount of memory. Basically, it brings the entire state information of all accounts and all slots into the main memory. This could be easily several 10s of GiB of data.

Reply:
I updated the visitor that it check each address immediately as spotted, furthermore the keys are checked by batches of 10.